phd-tests/tests/src/hyperv.rs

```diff
@@ -83,3 +89,186 @@ async fn hyperv_lifecycle_test(ctx: &Framework) {
+    cfg.guest_hv_interface(
+        propolis_client::types::GuestHypervisorInterface::HyperV {
+            features: vec![HyperVFeatureFlag::ReferenceTsc],
+        },
+    );
+    let mut vm = ctx.spawn_vm(&cfg, None).await?;
+    vm.launch().await?;
+    vm.wait_to_boot().await?;
+
+    #[derive(Debug)]
+    struct Reading {
+        taken_at: Instant,
+        guest_ns: u64,
+    }
+
+    impl Reading {
+        async fn take_from(vm: &TestVm) -> anyhow::Result<Self> {
+            let cmd =
+                "cat /proc/timer_list | grep \"now at\" | awk '{ print $3 }'";
+            let guest_ns = vm.run_shell_command(cmd).await?.parse::<u64>()?;
+
+            // It's important to minimize the time between the host and guest
+            // readings to avoid introducing skew when comparing sets of
+            // readings. Capturing the host timestamp after the guest time
+            // excludes the time needed to send bytes to the guest serial potr
+            // and wait for them to be echoed from this delta.
+            let taken_at = Instant::now();
+            Ok(Self { taken_at, guest_ns })
+        }
+
+        /// Compares `self` with an earlier reading, `other`, and returns the
+        /// difference between the measured elapsed time on the host and the
+        /// measured elapsed time on the guest, expressed as a percentage of the
+        /// measured elapsed time on the host.
+        fn compare_with_earlier(&self, other: &Reading) -> f64 {
+            let host_delta_ns =
+                i64::try_from((self.taken_at - other.taken_at).as_nanos())
+                    .expect("host delta is small enough to fit in an i64");
+
+            let guest_delta_ns = i64::try_from(self.guest_ns - other.guest_ns)
+                .expect("guest delta is small enough to fit in an i64");
+
+            let diff = (host_delta_ns - guest_delta_ns).unsigned_abs();
+            let diff_pct = (diff as f64) / (host_delta_ns as f64);
+
+            info!(
+                before = ?other,
+                after = ?self,
+                host_delta_ns,
+                guest_delta_ns,
+                diff,
+                diff_pct,
+                "compared time readings"
+            );
+
+            diff_pct
+        }
+    }
+
+    // The logic below takes pairs of time readings on the host and guest and
+    // compares them to check that the amount of elapsed time perceived by the
+    // guest is roughly equivalent to the "real" elapsed time on the host. The
+    // host and guest measurements can't be taken atomically, so some difference
+    // between these readings is expected (especially because the guest shell
+    // command needs to print its result to the serial port, and the harness has
+    // to read it).
+    //
+    // Accept a measurement error of up to 1% of the elapsed time between host
+    // readings. This is an extremely generous tolerance. It is chosen primarily
```

Comment:
Apparently it's not generous enough:

```
phd-runner: [TEST - EVENT] compared time readings
    after = Reading { taken_at: Instant { tv_sec: 1525, tv_nsec: 922518064 }, guest_ns: 2289389900 }
    before = Reading { taken_at: Instant { tv_sec: 1525, tv_nsec: 376576078 }, guest_ns: 1753546300 }
    diff = 10098386
    diff_pct = 0.018497177830173333
    file = phd-tests/tests/src/hyperv.rs
    guest_delta_ns = 535843600
    host_delta_ns = 545941986
    line = 208
    path = phd_tests::hyperv::hyperv_reference_tsc_elapsed_time_test
    target = phd_tests::hyperv
```

Reply:
This passed with flying colors on a rerun, though. Instead of loosening the tolerance, I might adjust the logic so that this part of the test fails if all, or a majority, of the samples exceed the tolerance, since that implies a lower likelihood that it's all just noise. Will mull this over a bit more (and other suggestions are especially welcome!).

Reply:
6330dff takes a swing at this.